pkg/api/resolver.go

```diff
@@ -0,0 +1,348 @@
+package api
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"time"
+
+	//graphql "github.com/graph-gophers/graphql-go"
+
+	"github.com/thrasher-redhat/internal-tools/pkg/db"
+	"github.com/thrasher-redhat/internal-tools/pkg/options"
+)
+
+var dateFormat = "2006-01-02"
+
+// Resolver represents the connection between the API and the data
+type Resolver struct {
+	dbClient db.Client
+	releases []options.Release
+}
+
+// NewResolver is a factory for Resolver
+func NewResolver(db db.Client, releases []options.Release) (*Resolver, error) {
+
+	// Add a release object for unspecified releases
+	noRelease := options.Release{
+		Name:    "---",
+		Targets: []string{"---"},
+	}
+
+	return &Resolver{
+		dbClient: db,
+		releases: append(releases, noRelease),
+	}, nil
+}
+
+// parseDatestamp accounts for empty date or special strings
+// TODO - whats the actual flow going to be for datestamps in regards to str vs time?
+func (r Resolver) parseDatestamp(datestamp string) (string, error) {
+	//var date time.Time
+	//var err error
+
+	// Default to latest
+	switch datestamp {
+	case "", "_latest":
+		date, err := r.dbClient.GetLatest()
+		if err != nil {
+			return "", err
+		}
+		datestamp = date.Format(dateFormat)
+	case "_earliest":
+		date, err := r.dbClient.GetEarliest()
+		if err != nil {
+			return "", err
+		}
+		datestamp = date.Format(dateFormat)
+	}
+	return datestamp, nil
+}
+
+// parseComponents checks for empty lists and works around the graphql
+// listpacker bug. Basically if you use a list that isn't []interface{}, then
+// Pack() uses a interface slice where the first element is given list
+// Current workaround is to use *[]type to get the actual array, so
+// dereference that pointer here to get the actual list.
+// https://github.com/graph-gophers/graphql-go/blob/b46637030579abd312c5eea21d36845b6e9e7ca4/internal/exec/packer/packer.go#L248-L250
+// TODO - version lock graphql-go for now as the fix will break this.
+func parseComponents(components *[]string) []string {
+	// NOTE - If we want to support "_all", we could check for it in keywords and return nil
+	// TODO - should we check that the given components actually exist?
```

Comment:
Soundboarding...

Not quite sure the proper path here.  We could silently fail when incorrect components are given - giving 0 results back for the "Pods" team (instead of the "Pod" team).  Or we could fail with a specific error.

Or we could check for the existence of such a component.  We would need to pull the list of components from the database and check there (adding another db query).  Theres also a rare corner case of a new or otherwise legitimate component in bugzilla that simply hasn't appeared in the database yet (perhaps component "foo" only has minor bugs that don't meet the predefined bugzilla search).  So querying for component "foo" would fail.  More importantly, querying for team "bar" that owns component "foo" (and others) would fail because "foo" isn't in the database.

So it doesn't look like checking does much for us and introduces a rather nasty corner case.  Going to fail silently as we will simply have no bugs for component "foo".